client-misk-jooq/src/main/kotlin/app/cash/backfila/client/misk/jooq/internal/OpenKeyRange.kt

```diff
@@ -0,0 +1,152 @@
+      .and(jooqBackfill.compareCompoundKey(upperBound) { keyCompare, compoundKeyValue ->
+        keyCompare.lte(compoundKeyValue)
+      })
+  }
+
+  fun betweenStartAndEndCondition(end: K): Condition {
+    return jooqBackfill.compareCompoundKey(start, startComparison)
+      .and(jooqBackfill.compareCompoundKey(end) { keyCompare, compoundKeyValue ->
+        keyCompare.lte(compoundKeyValue)
+      })
+  }
+
+  /**
+   * Builds a range that comes after the provided `end`.
+   */
+  fun nextRangeFor(end: K): OpenKeyRange<K> {
+    return OpenKeyRange(
+      jooqBackfill = jooqBackfill,
+      upperBound = upperBound,
+      start = end,
+      startComparison = { keyCompare: CompoundKeyComparer, compoundKeyValue: Record ->
+        keyCompare.gt(compoundKeyValue)
+      }
+    )
+  }
+
+
+  companion object {
+    /**
+     * Builds the initial range that we start with when iterating over a sequence of ranges.
+     */
+    fun <K> initialRangeFor(
+      jooqBackfill: JooqBackfill<K>,
+      request: GetNextBatchRangeRequest,
+      session: DSLContext
+    ): OpenKeyRange<K> {
+      val startComparison: CompoundKeyComparisonOperator =
+      // If this is the first batch, we want to start with the provided value on the backfila
+        // screen. If not, then, we need to start with one after the previous end value
+        if (request.previous_end_key == null) {
+          CompoundKeyComparisonOperator { keyComparer: CompoundKeyComparer, compoundKeyValue: Record ->
+            keyComparer.gte(compoundKeyValue)
+          }
+        } else {
+          CompoundKeyComparisonOperator { keyComparer: CompoundKeyComparer, compoundKeyValue: Record ->
+            keyComparer.gt(
+              compoundKeyValue
+            )
+          }
+        }
+
+      val start = request.previous_end_key
+        ?.let { jooqBackfill.fromByteString(request.backfill_range.start) }
+        ?: jooqBackfill.fromByteString(request.previous_end_key)
+
+      return OpenKeyRange(
+        jooqBackfill = jooqBackfill,
+        start = start,
+        startComparison = startComparison,
+        upperBound = computeUpperBound(
+          jooqBackfill, request, session,
+          jooqBackfill.compareCompoundKey(start, startComparison)
+        )
+      )
+    }
+
+    /**
+     * Get the upper bound of the scan range without applying any backfill filter conditions.
+     * This is so that we can restrict the upper bound when the filter is applied subsequently.
+     * If we don't restrict the range, then the query could perform badly.
+     *
+     * This SQL is essentially going to look like
+     * <pre>
+     * select <backfill fields>
+     *   from (
+     *      select <backfill fields>
+     *      from <backfill table>
+     *      where (backfill fields > end of previous range or >= start of backfill)
+     *      order by <backfill fields> asc
+     *      limit <scan size>
+ *      )
+     *  order by <backfill fields> desc
```

Comment:
this used to be max(), im guessing you had to switch to limit 1 because you want multiple fields as the key? have you tested it to work?

Reply:
This is based off the Jooq backfila implementation in the monorepo. I believe there are tests on this.

Those tests are probably worth moving in some form @Sethuraman

Reply:
Tests have been added.